### ISSUE TITLE
Git dependencies for semver-based matching: npm install not updating to new version

### ISSUE THREAD

**Issue by isurfbecause** [npm/npm#16760]
#### I'm opening this issue because:
  - [x ] npm is producing an incorrect install.

#### What's going wrong?
According to The npm Blog

> Git dependencies now support semver-based matching: npm install git://github.com/npm/npm#semver:^5 (#15308, #15666)

I was under the impression that the node modules installed via git dependencies would be updated after an npm install.

#### How can the CLI team reproduce the problem?
I have a project that installs `test-module` (v1.0.10) via github.
```
~/projects/npm5 (master)
 [144] → npm install test-module@https://github.com/isurfbecause/test-module/tarball/master#semver:^1
npm WARN npm5@1.0.0 No description
npm WARN npm5@1.0.0 No repository field.

added 1 package in 1.133s
```
npm5/node_modules/test-module/package.json
```

  "_from": "https://github.com/isurfbecause/test-module/tarball/master#semver:^1",
  "_id": "test-module@1.0.10",
  "_inBundle": false,
  "_integrity": "sha1-UPAzGJxcWMqi/glEjpiijyHV2bA=",
  "_location": "/test-module",
  "_phantomChildren": {},
  "_requested": {
    "type": "remote",
    "raw": "test-module@https://github.com/isurfbecause/test-module/tarball/master#semver:^1",
    "name": "test-module",
    "escapedName": "test-module",
    "rawSpec": "https://github.com/isurfbecause/test-module/tarball/master#semver:^1",
    "saveSpec": "https://github.com/isurfbecause/test-module/tarball/master#semver:^1",
    "fetchSpec": "https://github.com/isurfbecause/test-module/tarball/master#semver:^1"
  },
  "_requiredBy": [
    "#USER",
    "/"
  ],
  "_resolved": "https://github.com/isurfbecause/test-module/tarball/master#semver:^1",
  "_spec": "test-module@https://github.com/isurfbecause/test-module/tarball/master#semver:^1",
  "_where": "/Users/winstonyee/projects/npm5",
  "author": "",
  "bundleDependencies": false,
  "deprecated": false,
  "description": "",
  "license": "ISC",
  "main": "index.js",
  "name": "test-module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "version": "1.0.10"
}
```

Then I update the `test-module` packag.json
```
~/projects/test-module (master)
 [59] → npm version patch
v1.0.11
~/projects/test-module (master)
 [60] → git push --follow-tags
Counting objects: 4, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 400 bytes | 0 bytes/s, done.
Total 4 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
```
test-module/package.json

```
{
  "name": "test-module",
  "version": "1.0.11",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC"
}

```

Then in my project folder I will run

```
~/projects/npm5 (master)
 [146] → npm install
npm WARN npm5@1.0.0 No description
npm WARN npm5@1.0.0 No repository field.
```
However the npm5/node_modules/test-module/package.json hasn't changed and the version is still `1.0.10`

### supporting information:
```
npm -v prints
5.0.0
```
```
node -v
v6.9.1
```
 - `npm config get registry` prints:
```
npm config get registry
https://registry.npmjs.org/
```
 - macOS

**Comment by zkat** [comment]
This isn't being installed as a git dependency: `npm install test-module@https://github.com/isurfbecause/test-module/tarball/master#semver:^1` is a *remote tarball* installation. That means it's equivalent to saying `https://registry.npmjs.org/test-module/-/test-module-1.2.3.tgz`.

In order to use the semver range resolution feature, you need to actually go through git itself. This should work: `npm i 'github:isurfbecause/test-module#semver:^1'`. Give that a whirl and see if it fixes it for you :)

**Comment by isurfbecause** [comment]
When I run `npm i 'github:isurfbecause/test-module#semver:^1'` I get an error 

```
npm ERR! Cannot read property 'match' of null
```

Here is the debug.log

```
0 info it worked if it ends with ok
1 verbose cli [ '/Users/winstonyee/.nvm/versions/node/v6.9.1/bin/node',
1 verbose cli   '/Users/winstonyee/.nvm/versions/node/v6.9.1/bin/npm',
1 verbose cli   'i',
1 verbose cli   'github:isurfbecause/test-module#semver:^1' ]
2 info using npm@5.0.0
3 info using node@v6.9.1
4 silly install loadCurrentTree
5 silly install readLocalPackageData
6 silly fetchPackageMetaData error for github:isurfbecause/test-module#semver:^1 Cannot read property 'match' of null
7 verbose stack TypeError: Cannot read property 'match' of null
7 verbose stack     at resolve (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/pacote/lib/fetchers/git.js:136:29)
7 verbose stack     at plainManifest (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/pacote/lib/fetchers/git.js:106:10)
7 verbose stack     at BB.resolve.then.catch.catch.err (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/pacote/lib/fetchers/git.js:100:12)
7 verbose stack     at tryCatcher (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/util.js:16:23)
7 verbose stack     at Promise._settlePromiseFromHandler (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:512:31)
7 verbose stack     at Promise._settlePromise (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:569:18)
7 verbose stack     at Promise._settlePromise0 (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:614:10)
7 verbose stack     at Promise._settlePromises (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:689:18)
7 verbose stack     at Async._drainQueue (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/async.js:133:16)
7 verbose stack     at Async._drainQueues (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/async.js:143:10)
7 verbose stack     at Immediate.Async.drainQueues (/Users/winstonyee/.nvm/versions/node/v6.9.1/lib/node_modules/npm/node_modules/bluebird/js/release/async.js:17:14)
7 verbose stack     at runCallback (timers.js:637:20)
7 verbose stack     at tryOnImmediate (timers.js:610:5)
7 verbose stack     at processImmediate [as _immediateCallback] (timers.js:582:5)
8 verbose cwd /Users/winstonyee/projects/npm5
9 verbose Darwin 15.6.0
10 verbose argv "/Users/winstonyee/.nvm/versions/node/v6.9.1/bin/node" "/Users/winstonyee/.nvm/versions/node/v6.9.1/bin/npm" "i" "github:isurfbecause/test-module#semver:^1"
11 verbose node v6.9.1
12 verbose npm  v5.0.0
13 error Cannot read property 'match' of null
14 verbose exit [ 1, true ]
```

**Comment by zkat** [comment]
Aha, that one's more useful. I'll take a look :)

**Comment by zkat** [comment]
Oh this is a pretty easy fix. It looks like my git code didn't update to use `npm-package-arg`'s new semver parsing. Thanks for the report!

**Comment by zkat** [comment]
This will be fixed with the upcoming 5.0.1 patch. https://github.com/zkat/pacote/commit/77a676aec152e6d3654601c50188ba9130093800 fixed this. So I'm gonna close it.
